criu/cr-dump.c

```diff
@@ -204,6 +206,59 @@ static int collect_fds(pid_t pid, struct parasite_drain_fd **dfds)
 	if (!fd_dir)
 		return -1;
 
+	/* Before collecting fds, we need to bring io_uring to steady state,
+	 * since it can install fds into task's fdtable, and if we do it later,
+	 * during actual io_uring dump, we will miss dumping these files.
+	 */
+	while ((de = readdir(fd_dir))) {
+		if (dir_dots(de))
+			continue;
+
+		n = dirfd(fd_dir);
+		if (n == -1) {
+			close(pidfd);
+			return -1;
+		}
+
+		n = readlinkat(n, de->d_name, buf, sizeof(buf));
+		if (n == -1) {
+			close(pidfd);
+			return -1;
+		}
+
+		if (is_io_uring_link(buf)) {
+			if (!kdat.has_pidfd_open) {
+				pr_err("pidfd_open system call not supported\n");
+				return -ENOTSUP;
+			}
+
+			if (!kdat.has_pidfd_getfd) {
+				pr_err("pidfd_getfd system call not supported\n");
+				return -ENOTSUP;
+			}
+
+			if (pidfd == -1) {
+				pidfd = syscall(SYS_pidfd_open, pid, 0);
+				if (pidfd < 0) {
+					pr_err("Failed to open pidfd for pid %d\n", pid);
+					return pidfd;
+				}
+			}
+
+			if (io_uring_synchronize_fd(syscall(__NR_pidfd_getfd, pidfd, atoi(de->d_name), 0))) {
```

Comment:
criu/cr-dump.c:248:40: error: '__NR_pidfd_getfd' undeclared (first use in this function)